File: 1-typeorm-and-postgresql/backend/0-typescript/src/modules/cat/cat.service.ts
/**
 * Cat CRUD service — TypeORM repository + demo 1:1, 1:N, N:N relationships.
 */
import {
    Injectable, Logger, NotFoundException 
} from "@nestjs/common"
import {
    InjectRepository 
} from "@nestjs/typeorm"
import {
    Repository 
} from "typeorm"
import {
    Cat, Toy
} from "../../entities/postgresql/main"

/**
 * Cat Service — Business logic service for cats and relationships. Follows pattern: prepare → execute → confirm.
 */
@Injectable()
export class CatService {
    private readonly logger = new Logger(CatService.name)

    constructor(
    @InjectRepository(Cat)
    private readonly catRepository: Repository<Cat>,
    @InjectRepository(Toy)
    private readonly toyRepository: Repository<Toy>,
    ) {}

    /**
   * Retrieves all cats with their relationships (JOIN).
   *
   * @returns Promise<Cat[]> - List of cats with full info
   */
    async findAll(): Promise<Cat[]> {
        // [execute] Execute JOIN query with all related tables
        this.logger.log("Fetching all cats with relations...")
        return await this.catRepository.find({
            relations: ["passport",
                "toys",
                "owners"],
        })
    }

    /**
   * Creates a new cat with all relationships (CASCADE).
   *
   * @param catData - Cat data
   * @returns Promise<Cat> - Newly created cat
   */
    async create(catData: Partial<Cat>): Promise<Cat> {
        // [prepare] Prepare new instance from input data
        this.logger.log({
            message: "Preparing to create new cat", data: catData 
        })
        const cat = this.catRepository.create(catData)

        // [execute] Execute save to database (TypeORM auto-saves relations due to cascade)
        const savedCat = await this.catRepository.save(cat)

        // [confirm] Confirm success result
        this.logger.log({
            message: "Cat created successfully", id: savedCat.id 
        })
        return savedCat
    }

    /**
   * Returns details of a cat by ID.
   *
   * @param id - Cat ID
   * @returns Promise<Cat> - Detailed cat info
   */
    async findOne(id: number): Promise<Cat> {
        // [execute] Execute find cat with relationship JOIN
        const cat = await this.catRepository.findOne({
            where: {
                id 
            },
            relations: ["passport",
                "toys",
                "owners"],
        })

        // [confirm] Confirm existence
        if (!cat) {
            this.logger.error(`Cat with ID ${id} not found`)
            throw new NotFoundException(`Cat with ID ${id} not found`)
        }

        return cat
    }

    /**
     * Returns cat details by ID with explicit relation loading (passport, toys, owners).
     *
     * @param id - Cat ID
     * @returns Promise<Cat> - Cat + hydrated relations
     */
    async findWithRelations(id: number): Promise<Cat> {
        // [execute] Explicitly list relations to demonstrate eager-loading.
        const cat = await this.catRepository.findOne({
            where: {
                id,
            },
            relations: ["passport",
                "toys",
                "owners"],
        })

        // [confirm] Throw 404 if missing so the client can tell empty vs missing apart.
        if (!cat) {
            this.logger.error(`Cat with ID ${id} not found`)
            throw new NotFoundException(`Cat with ID ${id} not found`)
        }

        return cat
    }

    /**
     * Append a new Toy to an existing cat (demonstrates 1:N collection mutation).
     *
     * @param catId - Cat ID
     * @param toyData - New toy data
     * @returns Promise<Cat> - Cat after adding toy with refreshed collection
     */
    async addToy(catId: number, toyData: Partial<Toy>): Promise<Cat> {
        // [prepare] Make sure the cat exists before attaching a toy.
        const cat = await this.findOne(catId)

        // [execute] Create new Toy attached via the `cat` relation -- TypeORM auto-fills `catId`.
        const toy = this.toyRepository.create({
            ...toyData,
            cat,
        })
        await this.toyRepository.save(toy)

        // [confirm] Re-read cat with relations so the response reflects the freshest state.
        this.logger.log({
            message: "Toy added to cat", catId, toyId: toy.id,
        })
        return await this.findWithRelations(catId)
    }
}

File: 1-typeorm-and-postgresql/backend/1-java/src/main/java/academy/starci/typeormpostgresql/services/CatService.java
package academy.starci.typeormpostgresql.services;

import academy.starci.typeormpostgresql.entities.Cat;
import academy.starci.typeormpostgresql.entities.Toy;
import academy.starci.typeormpostgresql.exceptions.NotFoundException;
import academy.starci.typeormpostgresql.repositories.CatRepository;
import academy.starci.typeormpostgresql.repositories.ToyRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Business-logic service for the Cat domain.
 * <p>
 * Encapsulates all data-access operations and relation-wiring logic so the controller
 * stays thin (HTTP mapping only). Each method follows prepare → execute → return.
 * {@link Transactional} is applied to write operations so all DB changes commit or roll back together.
 */
@Service
public class CatService {

    /**
     * Repository for {@link Cat} — provides JPA CRUD and query-by-id.
     */
    private final CatRepository catRepository;

    /**
     * Repository for {@link Toy} — used when appending a single toy without touching the full cat graph.
     */
    private final ToyRepository toyRepository;

    /**
     * Constructor injection ensures dependencies are immutable after construction (no field injection).
     *
     * @param catRepository JPA repository for cats.
     * @param toyRepository JPA repository for toys.
     */
    public CatService(CatRepository catRepository, ToyRepository toyRepository) {
        this.catRepository = catRepository;
        this.toyRepository = toyRepository;
    }

    /**
     * Returns all cats. Spring Data JPA loads relations lazily by default;
     * the {@link CatRepository#findAll()} override in the repository uses
     * {@code @EntityGraph} or {@code JOIN FETCH} if eager loading is required.
     * For this demo, default lazy is acceptable because the serialisation context
     * is still open (Open-in-View enabled by default in Spring Boot).
     *
     * @return list of all cats (may be empty, never {@code null}).
     */
    public List<Cat> findAll() {
        return catRepository.findAll();
    }

    /**
     * Creates a new cat with all provided relations.
     * <p>
     * Hibernate requires the owning side of every relation to hold a back-reference
     * to the parent before calling {@code save}; otherwise the FK column is left null.
     * This method fixes the bidirectional links (passport.cat, toy.cat) before saving.
     *
     * @param cat cat data (deserialized from JSON — may contain nested passport, toys, owners).
     * @return the persisted cat with auto-generated IDs on all nested entities.
     */
    @Transactional
    public Cat create(Cat cat) {
        // Wire the inverse 1:1 side so Hibernate populates passport.cat back-reference.
        if (cat.getPassport() != null) {
            cat.getPassport().setCat(cat);
        }
        // Wire the owning FK side on each Toy so Hibernate writes catId correctly.
        if (cat.getToys() != null) {
            for (Toy toy : cat.getToys()) {
                toy.setCat(cat);
            }
        }
        // A single save cascades to passport (CascadeType.ALL), toys (CascadeType.ALL),
        // and owners (CascadeType.PERSIST + MERGE) via the Cat entity declarations.
        return catRepository.save(cat);
    }

    /**
     * Finds a cat by its primary key, throwing {@link NotFoundException} (HTTP 404) if absent.
     *
     * @param id the cat's primary key.
     * @return the found cat entity.
     * @throws NotFoundException if no cat with the given id exists.
     */
    public Cat findOne(Integer id) {
        // orElseThrow converts Optional.empty() into our domain exception (maps to HTTP 404 in controller).
        return catRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Cat with ID " + id + " not found"));
    }

    /**
     * Returns a cat with all relations explicitly loaded.
     * Delegates to {@link #findOne(Integer)} — Spring Data JPA's default fetch strategy
     * (EAGER on {@code @OneToOne}, LAZY on collections) means all relations are already
     * available within the Open-Session-In-View window.
     *
     * @param id the cat's primary key.
     * @return the cat with hydrated relations.
     */
    public Cat findWithRelations(Integer id) {
        return findOne(id);
    }

    /**
     * Appends a new toy to an existing cat (demonstrates 1:N collection mutation).
     * <p>
     * Rather than loading the full cat graph and re-saving, we create and persist the
     * Toy independently — only setting the FK via {@code toy.setCat(cat)}. This avoids
     * re-triggering cascades on the whole cat and is the recommended pattern for large collections.
     *
     * @param catId primary key of the cat that will own the new toy.
     * @param toy   toy data to persist (name is required).
     * @return the updated cat with the new toy included in the {@code toys} collection.
     * @throws NotFoundException if no cat with the given catId exists.
     */
    @Transactional
    public Cat addToy(Integer catId, Toy toy) {
        // Verify the cat exists before attaching the toy (fail fast with a clear 404).
        Cat cat = findOne(catId);

        // Link the toy to the cat — Hibernate uses this object reference to populate catId FK.
        toy.setCat(cat);
        toyRepository.save(toy);

        // Re-fetch the cat so the returned object reflects the new toy in the collection.
        return findOne(catId);
    }
}
